File: templates/sparse-tables/SparseGCD.cpp
// TEMPLATE BY ISHAANBUILDSTHINGS on github
// Construct a GCD sparse table
// O(n log n log(maxVal)) build time
// O(log(maxVal)) query range GCD, so effectively O(1)
// ⚠️ Not optimized, also my style is a bit weird as I populate sparse[power][left] even when it would go past the right edge, might be slower to build
#include <bits/stdc++.h>
using namespace std;
struct Sparse {
    int n;
    int LOG;
    vector<vector<int>> sparse; // sparse[power][left] is the GCD for that range
    Sparse(const vector<int>& arr) {
        n = arr.size();
        LOG = 32 - __builtin_clz(n);
        sparse.resize(LOG, vector<int>(n));
        for (int i = 0; i < n; i++) {
            sparse[0][i] = arr[i];
        }
        for (int power = 1; power < LOG; power++) {
            int width = 1 << power;
            int halfWidth = width / 2;
            for (int left = 0; left < n; left++) {
                int gcdLeft = sparse[power-1][left];
                int rightEdge = left + halfWidth;
                if (rightEdge < n) {
                    gcdLeft = gcd(gcdLeft, sparse[power-1][rightEdge]);
                }
                sparse[power][left] = gcdLeft;
            }
        }
    }
    int query(int l, int r) {
        int width = r - l + 1;
        int maxPow = 31 - __builtin_clz(width);
        int powWidth = 1 << maxPow;
        int left = sparse[maxPow][l];
        int right = sparse[maxPow][l + width - powWidth];
        return gcd(left, right);
    }
};

File: templates/sparse-tables/SparseGCD.py
# GCD sparse table
from math import gcd

class SparseGcd:
    def __init__(self, arr):
        self.n = len(arr)
        self.LOG = self.n.bit_length()
        self.sparse = [[0] * self.n for _ in range(self.LOG)]
        for i in range(self.n):
            self.sparse[0][i] = arr[i]
        for power in range(1, self.LOG):
            halfWidth = 1 << (power - 1)
            for left in range(self.n):
                val = self.sparse[power - 1][left]
                rightEdge = left + halfWidth
                if rightEdge < self.n:
                    val = gcd(val, self.sparse[power - 1][rightEdge])
                self.sparse[power][left] = val
    # inclusive [l, r]
    def query(self, l, r):
        width = r - l + 1
        maxPow = width.bit_length() - 1
        powWidth = 1 << maxPow
        return gcd(
            self.sparse[maxPow][l],
            self.sparse[maxPow][l + width - powWidth]
        )